router/proxy/reverseproxy.go

```diff
@@ -0,0 +1,332 @@
+var (
+	hopHeaders = []string{
+		"Te", // canonicalized version of "TE"
+		"Trailers",
+		"Transfer-Encoding",
+	}
+
+	serviceUnavailable = []byte("Service Unavailable\n")
+)
+
+// ReverseProxy is an HTTP Handler that takes an incoming request and
+// sends it to another server, proxying the response back to the
+// client.
+type ReverseProxy struct {
+	// The transport used to perform proxy requests.
+	transport *transport
+
+	// FlushInterval specifies the flush interval
+	// to flush to the client while copying the
+	// response body.
+	// If zero, no periodic flushing is done.
+	FlushInterval time.Duration
+
+	// ErrorLog specifies an optional logger for errors
+	// that occur when attempting to proxy the request.
+	// If nil, logging goes to os.Stderr via the log package's
+	// standard logger.
+	ErrorLog *log.Logger
+}
+
+// NewReverseProxy initializes a new ReverseProxy with an initial list of
+// backends, a stickyKey for encrypting sticky session cookies, and a flag
+// sticky to enable sticky sessions.
+func NewReverseProxy(bf BackendListFunc, stickyKey [32]byte, sticky bool) *ReverseProxy {
+	return &ReverseProxy{
+		transport: &transport{
+			getBackends:       bf,
+			stickyCookieKey:   stickyKey,
+			useStickySessions: sticky,
+		},
+		FlushInterval: 50 * time.Millisecond,
+	}
+}
+
+// ServeHTTP implements http.Handler.
+func (p *ReverseProxy) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	transport := p.transport
+	if transport == nil {
+		panic("router: nil transport for proxy")
+	}
+
+	outreq := prepareRequest(req)
+
+	if isConnectionUpgrade(req.Header) {
+		p.serveUpgrade(rw, outreq)
+		return
+	}
+
+	res, err := transport.RoundTrip(outreq)
+	if err != nil {
+		p.logf("router: proxy error: %v", err)
+		rw.WriteHeader(http.StatusServiceUnavailable)
+		rw.Write(serviceUnavailable)
+		return
+	}
+	defer res.Body.Close()
+
+	p.writeResponse(rw, res)
+}
+
+// ServeConn takes an inbound conn and proxies it to a backend.
+func (p *ReverseProxy) ServeConn(conn net.Conn) {
+	transport := p.transport
+	if transport == nil {
+		panic("router: nil transport for proxy")
+	}
+	defer conn.Close()
+
+	dconn, err := transport.Connect(conn.RemoteAddr())
+	if err != nil {
+		panic(err)
+		p.logf("router: proxy error: %v", err)
+		return
+	}
+
+	joinConns(conn, dconn)
+}
+
+func (p *ReverseProxy) serveUpgrade(rw http.ResponseWriter, req *http.Request) {
+	transport := p.transport
+	if transport == nil {
+		panic("router: nil transport for proxy")
+	}
+
+	res, uconn, err := transport.UpgradeHTTP(req)
+	if err != nil {
+		p.logf("router: proxy error: %v", err)
+		rw.WriteHeader(http.StatusServiceUnavailable)
+		rw.Write(serviceUnavailable)
+		return
+	}
+	defer uconn.Close()
+
+	p.writeResponse(rw, res)
+	res.Body.Close()
+
+	if res.StatusCode != 101 {
+		return
+	}
+
+	hj, ok := rw.(http.Hijacker)
+	if !ok {
+		p.logf("router: hijack failed")
+		return
+	}
+	dconn, bufrw, err := hj.Hijack()
```

Comment:
Drop the type check and do an unchecked assertion here, the panic will have detail and it should never happen at runtime if the use is correct.